homeassistant/components/trello/config_flow.py

```diff
@@ -0,0 +1,220 @@
+    """Handle a config flow for the initial setup of a Trello integration."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize config flow."""
+        self.api_key: str = ""
+        self.api_token: str = ""
+        self.user_email: str = ""
+        self.user_id: str = ""
+        self.ids_boards: dict[str, dict[str, str]] = {}
+        self.trello_adapter: TrelloAdapter
+
+    @staticmethod
+    @callback
+    def async_get_options_flow(
+        config_entry: config_entries.ConfigEntry,
+    ) -> config_entries.OptionsFlow:
+        """Create the options flow."""
+        return OptionsFlowHandler(config_entry)
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Prompt user for Trello API credentials.
+
+        :param user_input: None initially or users api_key and api_token.
+        :return:
+        """
+        if user_input is None:
+            return await self._show_creds_form()
+
+        self.api_key = user_input[CONF_API_KEY]
+        self.api_token = user_input[CONF_API_TOKEN]
+        self.trello_adapter = _create_trello_adapter(
+            user_input[CONF_API_KEY], user_input[CONF_API_TOKEN]
+        )
+
+        try:
+            member = await self._get_current_member()
+        except Unauthorized as ex:
+            return await self._show_error_creds_form(ex)
+
+        self.user_id = member.id
+        self.user_email = member.email
+        self.ids_boards = await self._fetch_all_boards()
+
+        return await self._show_board_form()
+
+    async def async_step_finish(self, user_input: dict[str, Any]) -> FlowResult:
+        """Select desired boards to have card counts of per list.
+
+        :param user_input: User's selected boards
+        """
+        boards: dict[str, dict] = await self._get_boards_lists(
+            user_input[USER_INPUT_BOARD_IDS]
+        )
+
+        await self.async_set_unique_id(self.user_id)
+        self._abort_if_unique_id_configured()
+
+        config_data: dict[str, str] = self._get_config_data()
+        config_options = {CONF_OPTIONS_BOARDS: boards}
+
+        return self.async_create_entry(
+            title=self.user_email, data=config_data, options=config_options
+        )
+
+    def _get_config_data(self) -> dict[str, str]:
+        return {
+            CONF_API_KEY: self.api_key,
+            CONF_API_TOKEN: self.api_token,
+            CONF_USER_ID: self.user_id,
+            CONF_USER_EMAIL: self.user_email,
+        }
+
+    async def _show_creds_form(self) -> FlowResult:
+        return self.async_show_form(
+            step_id="user", data_schema=CREDS_FORM_SCHEMA, last_step=False
+        )
+
+    async def _show_board_form(self) -> FlowResult:
+        return self.async_show_form(
+            step_id="finish",
+            data_schema=_get_board_select_schema(self.ids_boards),
+            last_step=True,
+        )
```

Comment:
I'd suggest getting rid of the separate functions for showing forms. It looks confusing for new contributors and you always have to trace back to the place where it was called. (for example what if this method is called while ids_boards is None for some reason. It can get confusing)

Reply:
I can see where you're coming from with having to jump back and forth. I think it's more important to keep the `async_step` methods as readable as possible though. That's where the high level logic is and all the form definitions makes it much more difficult to grok. Also having all the form definitions encapsulated in these methods makes it easier to reference.

As a new contributor myself, it was really difficult to read through a lot of config flows for reference because they didn't abstract away a lot of higher level behavior.

I'm with you there on the `self.ids_boards`  reference though. To make it easier to reason about, I'm going to pass that in as a param.

Reply:
I agree with you, in one of my first PRs I also did this and then I got this as feedback (https://github.com/home-assistant/core/pull/91601#discussion_r1169927319)

Reply:
Thanks, that's good to know. I still feel pretty strongly that this should be abstracted away into a method for readability in the `async_step` method. 

Otherwise, it does seem like `add_suggested_values_to_schema` would be appropriate to use here. However, looking at it again, I actually don't need to set defaults anymore since I've removed the options flow. So I'm removing the logic that would make `add_suggested_values_to_schema` necessary, but I'll definitely implement it in the follow up PR.